osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -220,8 +220,30 @@ private void beatmapSetsChanged(IRealmCollection<BeatmapSetInfo> sender, ChangeS
 
             if (changes == null)
             {
-                // initial load
-                loadBeatmapSets(sender);
+                // During initial population, we must manually account for the fact that our original query was done on an async thread.
+                // Since then, there may have been imports or deletions.
+                // Here we manually catch up on any changes.
+                var populatedSets = new HashSet<Guid>();
+                foreach (var s in beatmapSets)
+                    populatedSets.Add(s.BeatmapSet.ID);
+
+                var realmSets = new HashSet<Guid>();
+                foreach (var s in sender)
```

Comment:
@peppy sorry for commenting on a merged PR, but I noticed in the perf trace foreaching the result has a non-trivial perf cost due to snapshotting it. While we still recommend using foreach for most use cases, since you don't mutate the collection, you should be able to use a normal for loop and access the elements by index. I'd recommend testing it, but my guess is that you should shave off the ~150ms `GetEnumerator` cost.

Reply:
Sounds good, thanks for the heads up.